### PR TITLE
costa: fix upload tarball error

### PR DESCRIPTION
**Pull request by FeelyChau**
Fix bug:
https://github.com/alibaba/pipcook/issues/445
### Test code:
```js
var fs = require('fs-extra');

async function test() {
  return new Promise((resolve, reject) => {
    const fr = fs.createReadStream('/path/to/read');
    const fw = fs.createWriteStream('/path/to/write');
    fr.pipe(fw);
    fr.on('end', () => {
      console.log('read end');
      resolve();
    });
    fw.on('finish', () => {
      console.log('write finish');
    });
  })
}

test().then(() => {
  console.log('test end');
});
```
### output:
* case A:
```shell
read end
write finish
test end
```
* case B:
```
read end
test end
write finish
```

If we get the case B, the error happens.

### CODE REVIEW

**Comment by yorkie** [comment]
Why not add the case in the codebase?

File: packages/costa/src/runtime.ts [hunk 1]
```diff
@@ -292,13 +302,9 @@ export class CostaRuntime {
...
     const filename = path.join(fileDir, 'pkg.tgz');
+    const writeStream = createWriteStream(filename);
     await mkdirp(fileDir);
-    await new Promise((resolve, reject) => {
-      const writeStream = createWriteStream(filename);
-      stream.pipe(writeStream);
-      stream.on('end', resolve);
-      stream.on('error', reject);
-    });
+    await this.saveFile(stream, writeStream);
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:307]: Could you please move the writeStream to saveFile, too?

Comment (FeelyChau) [reply packages/costa/src/runtime.ts:307]: If that, the unit test will be unable.

Comment (FeelyChau) [reply packages/costa/src/runtime.ts]: Moved this function to `utils.ts`.

File: packages/costa/src/utils.ts [hunk 2]
```diff
@@ -17,3 +17,17 @@ export function pipe(readable: NodeJS.ReadableStream, writable: NodeJS.WritableS
     writable.write(data);
   });
 }
+
+/**
+ * pipe read stream to write steam with finish event
+ * @param readStream the read stream
+ * @param writeStream the write stream
+ */
+export function pipeWithFinish(readStream: NodeJS.ReadableStream, writeStream: NodeJS.WritableStream): Promise<void> {
```
Comment (yorkie) [inline packages/costa/src/utils.ts:26]: The name looks wired a bit, `finish` is just an event.

Comment (FeelyChau) [reply packages/costa/src/utils.ts:26]: How about `pipePromisify`?

File: packages/costa/src/runtime.ts [hunk 3]
```diff
@@ -81,22 +81,20 @@ function extractPackageJsonFromReadable(readable: NodeJS.ReadableStream, pkgFile
...
-    extract.on('error', reject);
     extract.on('entry', (header, stream, next) => {
       if (header.name === pkgFilename) {
         stream.on('data', (buf) => packageJson += buf);
       }
       stream.once('end', next);
       stream.resume();
     });
-    extract.on('finish', () => {
+    pipeWithFinish(readable, extract).then(() => {
```
Comment (yorkie) [inline packages/costa/src/runtime.ts:91]: No necessary to return a Promise here, we could use a callback style and call promisify for async/await use-case.

File: packages/costa/src/utils.ts (hunk 2 above)
Comment (yorkie) [reply packages/costa/src/utils.ts:26]: I think `safePipe` is more specific.

Comment (yorkie) [reply packages/costa/src/utils.ts:26]: Or `pipeGracefully`?

File: packages/costa/src/runtime.ts (hunk 3 above)
Comment (FeelyChau) [reply packages/costa/src/runtime.ts:91]: This part of the code could also be simplified a little bit.

Comment (FeelyChau) [reply packages/costa/src/runtime.ts]: Updated.

File: packages/costa/src/utils.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/costa/src/utils.ts]: pipeGracefully +1

**Pull request merged**